software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+#!/usr/bin/python3
+
+import argparse, re
+
+"""
+BSG Manycore linker command file generator.
+
+Usage: ./bsg_manycore_link_gen.py -h
+
+Bandhav Veluri
+02/11/2020
+"""
+
+class bsg_manycore_link_gen:
+  """
+  A configurable linker command file generator for linking BSG Manycore
+  kernels. Kernel memory has private and shared regions backed by 4KB
+  on-tile sram (DMEM) and off-chip DRAM, respectively. Configuration options
+  include default location of data (dmem or dram), stack pointer and
+  dram memory size. These options won't affect the following:
+
+  1. Data attributed to .dmem* and .dram* sections. Following decalaration
+
+     int foo __attribute__((section (".dram")));
+
+     would place foo in dram region irrespective of default data location.
+
+  2. Data in bsg_manycore_lib would always be placed in dmem region.
+
+  
+  Note on program memory:
+  
+  Manycore's PC width is 24-bit, addressing a total of 16MB program space.
+  Since the 31:25 MBSs are assumed to be 0s, linking should also assume
+  0x0-0x01000000 as the valid program space. This means that .text section VMAs
+  lie in 0x0-0x01000000. On the other hand, .text section is loaded to DRAM and
+  as result, it's LMAs shouls be >0x80000000. So, .text section is loaded in the
+  region 0x80000000-0x81000000. Consequently, VMA's 0x80000000-0x81000000 are
+  are reserved for .text section as well.
+  """
+
+  _opening_comment = \
+    "/**********************************\n" \
+    + " BSG Manycore Linker Script \n\n"
+
+  def __init__(self, default_data_loc, dram_size, sp):
+    self._default_data_loc = default_data_loc
+    self._dram_size = dram_size
+    self._sp = sp
+    self._opening_comment += \
+        " data default: {0}\n".format(default_data_loc) \
+        + " dram memory size: 0x{0:08x}\n".format(dram_size) \
+      + "***********************************/\n"
+
+  # Format:
+  # <mem region name> : [<access>, <start address>, <size>]
+  def _memory_regions(self, mem_regions):
+    script = 'MEMORY {\n'
+
+    for m, access, origin, size in mem_regions:
+      script += "{0} ({1}) : ORIGIN = 0x{2:08x}, LENGTH = 0x{3:08x}\n".format(
+          m, access, origin, size)
+
+    script += '}\n'
+
+    return script
+
+  def _section(self, name, address, vma_region, lma_region,
+      in_sections, in_objects):
```

Comment:
You can take this feedback as a general FYI ("Hey this is another coding style") or you can address it if you feel like it.

From my reading it seems like the _section and _memory_region methods are prime candidates to be split into objects. For example, each section has attributes, specific arguments, and a string representation. Furthermore, there's specific checking that can be done for each section. Therefore, I conceptually think of each section as it's own Object (even though

For a simple python script (like this) it's not necessary, but in large programs it can break it up, modularize, and create abstractions. Take it as you will

Reply:
I had thought about this quite a bit when writing this and felt that _section and _memory_region doesn't need any state. So making them plain function that gives one output for each input seemed right.

Also, my primary consideration with this script was if someone want change the script, they most probably should only have to touch `.script()` method. So, `.script(..)` method should feel more or less like writing actual `.ld` script with the additional power of Python. So didn't want to deviate from `.ld` semantics by too much "class"ification ;).